### ISSUE TITLE
nested: NumberFormatException[Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]

### ISSUE THREAD

**Issue by hero007asd** [elasticsearch/elasticsearch#9191]
I create the mappings and a snippet as following:
>"weight": {
    "type": "integer"
}

And ,when i search by 
>{
  "sort": [
    {
      "weight": {
        "order": "desc"
      }
    },
    "_score"
  ]
}

es server reports:
>nested: NumberFormatException[Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]

I'm using es-1.4.2 on centOs.

**Comment by Hoannv** [comment]
I had same issue for yesterday

```
school:
                    mappings:
                        name: ~
                        shortName: ~
                        city: ~
                        address: ~
                        description: ~
                        rankingTotal: { type : integer }
```
Where I order by rankingTotal, it throws below error.
```
SearchPhaseExecutionException[Failed to execute phase [query_fetch], all shards failed; shardFailures {[l2T3o59-S5GXM3CRgQlomw][abcxyz][0]: QueryPhaseExecutionException[[abcxyz][0]: query[filtered(+((name:boston^4.0 | shortName:boston | city:boston^2.0 | address:boston | description:boston | area.title:boston | area.shortTitle:boston | area.parent.title:boston | area.parent.shortTitle:boston) ToParentBlockJoinQuery (filtered((topSchools.speciality.title:boston | topSchools.speciality.prefixTitle:boston | topSchools.speciality.suffixTitle:boston | topSchools.speciality.shortTitle:boston | topSchools.speciality.shortDescription:boston))->cache(_type:__topSchools)) ToParentBlockJoinQuery (filtered(profiles.content:boston)->cache(_type:__profiles))))->cache(_type:school)],from[0],size[10],sort[,,]: Query Failed [Failed to execute main query]]; nested: ElasticsearchException[java.lang.NumberFormatException: Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; nested: UncheckedExecutionException[java.lang.NumberFormatException: Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; nested: NumberFormatException[Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; }]
```
But it's working fine when I remove order by rankingtotal

**Comment by rjernst** [comment]
@hero007asd or @Hoannv can you give more details about your setup? It would be helpful to see the query you are running, and perhaps a couple simplified documents that will reproduce the issue?

**Comment by Hoannv** [comment]
I attached one simplified document which is returned after query.

Query has sort by three fields as following:
```
  'sort' => 
    array (size=3)
      'rankingExclude' => 
        array (size=1)
          'order' => string 'asc' (length=3)
      '_score' => 
        array (size=1)
          'order' => string 'desc' (length=4)
      'rankingTotal' => 
        array (size=1)
          'order' => string 'asc' (length=3)
```
The following is one simplified document

```
    array (size=6)
      '_index' => string 'findmba' (length=7)
      '_type' => string 'school' (length=6)
      '_id' => string '1017' (length=4)
      '_score' => float 3.737905
      '_source' => 
        array (size=1)
          'rankingTotal' => int 108
      'sort' => 
        array (size=3)
          0 => string 'false' (length=5)
          1 => float 3.737905
          2 => 994323212134343423
```
I don't know why: 
- document  has 'rankingTotal' => int 108
- but the third parameter of 'sort by rankingTotal' is a long value '2 => 994323212134343423'

When I removed sort by rankingTotal, it's working properly.

**Comment by rjernst** [comment]
@Hoannv I still don't understand.  Could you give me:
1. Example documents (the full source, as json, not within query results)?
2. The actual query you are running (preferrable as json)?

**Comment by Hoannv** [comment]
I'm using FOSElasticaBundle for my project. Here are what you asked.

Example document:

```
{"_index":"mba","_type":"school","_id":"2","_version":32,"found":true,"_source":{"name":"Samford University - Brock School of Business","shortName":"Samford - Brock","city":"Birmingham","address":"800 Lakeshore Drive","description":null,"rankingTotal":1042,"rankingExclude":false,"rankings":[],"accreditations":[{"id":1}],"area":{"id":16,"title":"Alabama","shortTitle":"AL","parent":{"id":15,"title":"United States","shortTitle":"USA"}},"programs":[{"languages":[{"id":1}],"averageGmat":0,"concentrations":[{"id":1},{"id":14}],"programType":{"id":13}},{"languages":[{"id":1}],"averageGmat":0,"concentrations":[{"id":14},{"id":19}],"programType":{"id":13}},{"languages":[{"id":1}],"averageGmat":0,"concentrations":[{"id":14}],"programType":{"id":13}},{"languages":[{"id":1}],"averageGmat":0,"concentrations":[{"id":1},{"id":12},{"id":14},{"id":21},{"id":33},{"id":36}],"programType":{"id":2}},{"languages":[{"id":1}],"averageGmat":null,"concentrations":[{"id":14},{"id":15}],"programType":{"id":13}}],"topSchools":[],"profiles":[]}}
```

And query
```
{"sort":{"rankingExclude":{"order":"asc"},"_score":{"order":"desc"},"rankingTotal":{"order":"asc"}},"query":{"bool":{"must":[{"bool":{"should":[{"query_string":{"query":"Boston","fields":["name^4","shortName","city^2","address","description","area.title","area.shortTitle","area.parent.title","area.parent.shortTitle"],"default_operator":"AND"}},{"nested":{"path":"topSchools","query":{"query_string":{"query":"Boston","fields":["topSchools.speciality.title","topSchools.speciality.prefixTitle","topSchools.speciality.suffixTitle","topSchools.speciality.shortTitle","topSchools.speciality.shortDescription"],"default_operator":"AND"}}}},{"nested":{"path":"profiles","query":{"query_string":{"query":"Boston","fields":["profiles.content"],"default_operator":"AND"}}}}]}}]}}}
```

**Comment by clintongormley** [comment]
Hi @hero007asd 

Do you have another field called `rankingTotal` in a different type in the same index? Also, do you create your mappings explicitly or are you relying on dynamic mapping?

Either way, it sounds like you have two versions of the `rankingTotal` field which are mapped differently.  When you try to load the fielddata to sort on this field, it expects one datatype, then suddenly encounters a different datatype, hence the exception.

This will be fixed by https://github.com/elasticsearch/elasticsearch/issues/8870

**Comment by Hoannv** [comment]
@clintongormley: no, there is only one rankingTotal in mapping

**Comment by clintongormley** [comment]
@hero007asd and do you rely on dynamic mapping or do you set the mapping specifically?

**Comment by dibeesh** [comment]
@hero007asd , @clintongormley @rjernst Any update on this. Am also facing the same problem when sorting an integer parameter Elastic search 1.3.7
Mapping 
{
   "outboxprov1": {
      "mappings": {
         "deleted_user": {
            "properties": {
               "status": {
                  "type": "integer"
               }
            }
         }
      }
   }
}
Query 
POST /outboxprov1/deleted_user/_search
{
  "sort": [
    {
      "status": {
        "order": "desc"
      }
    }
  ], 
   "query": {
      "match_all": {
      }
   }
}
Error 


{
   "error": "SearchPhaseExecutionException[Failed to execute phase [query], all shards failed; shardFailures {[0kPf5NYmRlC15ov89wjQ2g][outboxprov1][0]: QueryPhaseExecutionException[[outboxprov1][0]: query[ConstantScore(cache(_type:user))],from[0],size[10],sort[<custom:\"user_type\": org.elasticsearch.index.fielddata.fieldcomparator.LongValuesComparatorSource@1e9a2f49>!]: Query Failed [Failed to execute main query]]; nested: ElasticsearchException[java.lang.NumberFormatException: Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; nested: UncheckedExecutionException[java.lang.NumberFormatException: Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; nested: NumberFormatException[Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; }{[0kPf5NYmRlC15ov89wjQ2g][outboxprov1][1]: QueryPhaseExecutionException[[outboxprov1][1]: query[ConstantScore(cache(_type:user))],from[0],size[10],sort[<custom:\"user_type\": org.elasticsearch.index.fielddata.fieldcomparator.LongValuesComparatorSource@2e5a1ec5>!]: Query Failed [Failed to execute main query]]; nested: ElasticsearchException[java.lang.NumberFormatException: Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; nested: UncheckedExecutionException[java.lang.NumberFormatException: Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; nested: NumberFormatException[Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; }{[0kPf5NYmRlC15ov89wjQ2g][outboxprov1][2]: QueryPhaseExecutionException[[outboxprov1][2]: query[ConstantScore(cache(_type:user))],from[0],size[10],sort[<custom:\"user_type\": org.elasticsearch.index.fielddata.fieldcomparator.LongValuesComparatorSource@4e1d82e2>!]: Query Failed [Failed to execute main query]]; nested: ElasticsearchException[java.lang.NumberFormatException: Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; nested: UncheckedExecutionException[java.lang.NumberFormatException: Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; nested: NumberFormatException[Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; }{[0kPf5NYmRlC15ov89wjQ2g][outboxprov1][3]: QueryPhaseExecutionException[[outboxprov1][3]: query[ConstantScore(cache(_type:user))],from[0],size[10],sort[<custom:\"user_type\": org.elasticsearch.index.fielddata.fieldcomparator.LongValuesComparatorSource@7c2ff1fb>!]: Query Failed [Failed to execute main query]]; nested: ElasticsearchException[java.lang.NumberFormatException: Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; nested: UncheckedExecutionException[java.lang.NumberFormatException: Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; nested: NumberFormatException[Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; }{[0kPf5NYmRlC15ov89wjQ2g][outboxprov1][4]: QueryPhaseExecutionException[[outboxprov1][4]: query[ConstantScore(cache(_type:user))],from[0],size[10],sort[<custom:\"user_type\": org.elasticsearch.index.fielddata.fieldcomparator.LongValuesComparatorSource@6c31a161>!]: Query Failed [Failed to execute main query]]; nested: ElasticsearchException[java.lang.NumberFormatException: Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; nested: UncheckedExecutionException[java.lang.NumberFormatException: Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; nested: NumberFormatException[Invalid shift value in prefixCoded bytes (is encoded value really an INT?)]; }]",
   "status": 500
}

**Comment by clintongormley** [comment]
@dibeesh @Hoannv this is likely caused by relying on dynamic mapping, and a field is being added with different mappings on different shards.  When the shard with the "wrong" mapping is moved to another node, the official mapping is applied, which doesn't reflect the data written to disk.

You can work around this by creating your mappings manually before indexing.  This issue will be resolved by #8688
